Time stretching 3 of 6 repair drawing tools for stretched audio

saintmatthieu opened the pull request:
Resolves: #4850

This PR repairs drawing operations on stretched clips.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

## QA
- [x] editing one sample at a time by clicking with pencil,
- [x] editing many samples by dragging pencil across clips, some stretched, others not, some mono, others stereo.
- [x] same exercise with brush tool (select pencil then press alt)
- [x] trying to draw samples when individual samples are not visible (because zoomed out) issues the usual error message.
- [x] (Paul) test all in the pencil tool mode and again in multi-tool mode
- [x] (Paul) hit-testing on samples in multi tool with nontrivial envelope (this doesn’t work quite right in last release, but be sure it is not worse)
- [x] (Paul) include paintbrush cases where there is hidden smart clip data, and where clips (maybe with hidden data, maybe with different stretches) are touching or nearly touching
- [x] (Paul) I think we're also fixing an old misbehavior:  click and drag with the pencil sometimes did nothing but made a vacuous undo/redo item
- [x] (Paul) also remember (I had forgot!) that a drag with Ctrl key down adjusts only one pixel
- [x] Clicking or dragging edits nearest sample (not previous or next - nearest)
- [x] Drag-drawing on a clip's rectangle (not out-of-bound) always updates the nearest sample, not just when the pencil crosses its x-location
- [x] Works as well for stereo

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -371,6 +372,74 @@ class WAVE_TRACK_API WaveTrack final
+   /*!
+    * @brief Gets as many samples as it can, but no more than `2 *
+    * numSideSamples + 1`, centered around `t`. Reads nothing if
+    * `GetClipAtTime(t) == nullptr`. Useful to acess samples across clip
```
“Access”

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -371,6 +372,74 @@ class WAVE_TRACK_API WaveTrack final
+
+   /*!
+    * @brief Similar to GetFloatsFromTime, but for writing. Sets as many samples
+    * as it can according to the same ruls as GetFloatsFromTime. Leaves the
```
“Rules”

saintmatthieu commented:
This morning I came to remember maybe my very first experience of Audacity. It was at the National University of Ireland, Maynooth, in the computer music introductory module. I remember the lecturer saying "and you can even draw the samples with the pencil". I must have been impressed to still remember it - it was in 2005-2006. Great educational tool. @rbdannenberg @Paul-Licameli

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.h:
```diff
@@ -309,6 +318,20 @@ class WAVE_TRACK_API WaveClip final :
+   bool
+   GetFloatAtTime(double t, size_t iChannel, float& value, bool mayThrow) const;
+
+   void SetFloatsFromTime(
```
Please comment that the new setting functions will succeed with out-of-bounds requests, only changing what is in bounds.

Paul-Licameli replied:
`GetSamples` (not a new function) should also comment that `start` is relative to play start
`GetFloatAtTime` should comment that `t` is relative to play start

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -371,6 +372,74 @@ class WAVE_TRACK_API WaveTrack final
+    * numSideSamples + 1`, centered around `t`. Reads nothing if
+    * `GetClipAtTime(t) == nullptr`. Useful to acess samples across clip
+    * boundaries, as it spreads the read to adjacent clips, i.e., not separated
+    * by silence from clip at `t`.
```
Is this description really what we want?

To be REALLY strictly conservative of program behavior, in the case of not time stretching, it's possible two clips were separated by a tiny gap of one or two sample intervals, but the paint brush might affect both.

And maybe nobody in the history of Audacity actually ever did this.

saintmatthieu replied:
Now that clips can be stretched, we have to choose between constant time or constant sample-count brush radius. The approach I took is the latter. But then, samples of a clip 10 seconds away should not be taken into account. So I went for the simple rule that they have to touch each other, or nothing happens.
I also reckon we won't run into trouble because of this subtle change of behaviour.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -371,6 +372,74 @@ class WAVE_TRACK_API WaveTrack final
+   size_t GetFloatsFromTime(
+      double t, size_t iChannel, float* buffer, size_t numSamples,
+      bool mayThrow,
+      PlaybackDirection direction = PlaybackDirection::backward) const;
```
`backward` seems surprising as the default value.  But you never call it with the default.

saintmatthieu replied:
Right, makes no particular sense - removed it (and moved this method to `private`).

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -371,6 +372,74 @@ class WAVE_TRACK_API WaveTrack final
+   void SetFloatsFromTime(
+      double t, size_t iChannel, const float* buffer, size_t numSamples,
+      sampleFormat effectiveFormat,
+      PlaybackDirection direction = PlaybackDirection::backward);
```
ditto, default argument

Paul-Licameli commented on src/tracks/playabletrack/wavetrack/ui/SampleHandle.cpp:
```diff
@@ -93,8 +99,12 @@ namespace {
+      const auto clip = wt->GetClipAtTime(time);
+      if (!clip)
+         // Don't bother the user about that with a pop-up.
+         return true;
```
This function is used in two places, and this change, while avoiding that pop-up when you click between clips, also will change the hit test to return a handle in the case that it did not do so before.

Does clicking and dragging in the empty space between clips make any vacuous undo/redo items?

Paul-Licameli replied:
Aha, I tried the build.  Yes it does!

So instead the hit testing function should also be changed to return null when there is no clip.

Paul-Licameli replied:
Actually... that's an old defect, which you might fix.

I just tried released Audacity.  When the magnification does show samples, but you click and drag outside of a clip, still you do get a bogus undo/redo item.

Paul-Licameli replied:
Actually an old problem, and I was confused about something.  `HitTest` only happens with the * tool button.  With the pencil tool button, you can click anywhere and it has always been like that.

Paul-Licameli replied:
`SampleHandle::HitAnywhere` is also used in WaveformView.cpp in this case.

saintmatthieu replied:
> Actually... that's an old defect, which you might fix.

Each time `SampleHandle::Release` is called (when letting go of the mouse with pencil tool activated), a history item is pushed. The easiest solution to this I can think of would be to create a `bool mSampleMoved`, whose value is set to true whenever a sample is modified, checked when releasing, and then set to false again. That'd be something totally new, and not completely trivial to review.

Paul-Licameli commented on src/tracks/playabletrack/wavetrack/ui/CutlineHandle.cpp:
```diff
@@ -169,7 +169,16 @@ UIHandle::Result CutlineHandle::Click
-      else if (mLocation.typ == WaveTrackLocation::locationMergePoint) {
+      // todo(mhodgkinson) Bluntly disabling this Join shortcut for now. There
+      // exists an issue that wants to do this properly:
+      // https://github.com/audacity/audacity/issues/5020
```
I think that's not the right issue number

Paul-Licameli commented:
> This morning I came to remember maybe my very first experience of Audacity. It was at the National University of Ireland, Maynooth, in the computer music introductory module. I remember the lecturer saying "and you can even draw the samples with the pencil". I must have been impressed to still remember it - it was in 2005-2006. Great educational tool. @rbdannenberg @Paul-Licameli

Would you believe I hadn’t heard of, less used it, before December 2012?

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.h:
```diff
@@ -189,9 +189,18 @@ class WAVE_TRACK_API WaveClip final :
-   // todo comment
+   /*!
+    * Accessors for visible boundaries within a track after stretching, that is
+    * (ommitting rounding):
```
Omitting

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -455,6 +455,74 @@ class WAVE_TRACK_API WaveTrack final
+    * as it can according to the same ruls as GetFloatsFromTime. Leaves the
+    * other samples untouched. @see GetFloatsFromTime
+    */
+   void SetFloatsFromTime(
```
You need more comments to clarify:

When direction is backward, the buffer of given samples is not reversed, and `t` rounds to the time of the clip sample corresponding to the past-the-end position.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -455,6 +455,74 @@ class WAVE_TRACK_API WaveTrack final
+    *
+    * @return The number of samples actually copied.
+    */
+   size_t GetFloatsFromTime(
```
Clarify that the resulting samples are not reversed when direction is backward.

saintmatthieu replied:
Moved those to `private` and made comments a bit more precise.

saintmatthieu commented:
@Paul-Licameli just pushed [05c4f37](https://github.com/audacity/audacity/pull/5042/commits/05c4f37a53ec2ebe42ddfbfd2b478d4cd1e6eb01), addressing all your comments.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -884,6 +931,26 @@ class WAVE_TRACK_API WaveTrack final
 
+   /*!
+    * @brief Helper for GetFloatsCenteredAroundTime. If `direction ==
+    * PlaybackDirection::Backward`, fetches samples to the left of `t`, exluding
```
"excluding"

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -197,6 +197,52 @@ void WaveClip::SetSamples(size_t ii,
+   const auto effectiveStart = std::max(sampleCount { 0 }, maybeNegativeStart);
+   const auto effectiveEnd =
+      std::min(GetVisibleSampleCount(), maybeOutOfBoundEnd);
+   const auto effectiveLen = (effectiveEnd - effectiveStart).as_size_t();
```
As always, I see `as_size_t` and think, what's the proof of satsifaction of the assertion?

Here, "it's easy to prove that the expression is bounded by numFloats".  So just say that in a comment.

saintmatthieu replied:
Ok.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -197,6 +197,52 @@ void WaveClip::SetSamples(size_t ii,
+   const auto effectiveLen = (effectiveEnd - effectiveStart).as_size_t();
+   if (effectiveLen == 0)
+      return;
+   const auto numLeadingZeros = (effectiveStart - maybeNegativeStart).as_size_t();
```
likewise

Paul-Licameli replied:
See latest in the conversation

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -197,6 +197,52 @@ void WaveClip::SetSamples(size_t ii,
+      reinterpret_cast<const char*>(buffer + numLeadingZeros);
+   SetSamples(
+      iChannel, offsetBuffer, floatSample, effectiveStart, effectiveLen,
+      effectiveFormat);
```
Please comment in the header that the function may rewrite only a subrange of `[buffer, buffer + nFloats)`, leaving unchanged the buffer positions not corresponding to part of the clip play region.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -197,6 +197,52 @@ void WaveClip::SetSamples(size_t ii,
+{
+   SetFloatsFromTime(
+      t - SamplesToTime(numSideSamples), iChannel, buffer,
+      2 * numSideSamples + 1, effectiveFormat);
```
Please comment that `[buffer, buffer + 2 * numSizeSamples + 1)` is assumed to be a valid span of addresses, and likewise that only a subrange may be overwritten.

In other words zero-filling is not promised for the rest.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -477,6 +477,53 @@ class WAVE_TRACK_API WaveTrack final
+   void SetFloatsWithinTimeRange(
+      double t0, double t1, size_t iChannel,
+      const std::function<float(double sampleTime)>& producer,
+      sampleFormat efffectiveFormat);
```
"effective"

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+   if (!clip)
+      return;
+   t = clip->SamplesToTime(clip->TimeToSamples(t - clip->GetPlayStartTime())) +
+       clip->GetPlayStartTime();
```
`WaveClip::SnapToSample` analogous to the `WideSampleSequence` member function might be nice

saintmatthieu replied:
It might easily be confused for `WaveTrack::SnapToSample`, though. This manipulation is safe here in this `RoundToNearestClipSample` function, I'd leave it as it is if there aren't other needs elsewhere for it.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+struct BufferCharType<const float*>
+{
+   using type = constSamplePtr;
+};
```
I might simplify all that with 
`#include <type_traits>`
and
`template<typename T> using BufferCharType =
   std::conditional_t<std::is_const_v<std::remove_pointer_t<T>>, constSamplePtr, samplePtr>;`

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+   const auto forward = direction == PlaybackDirection::forward;
+   while (clip)
+   {
+      const auto args = GetSampleAccessArgs<float*>(
```
I think the template argument can be omitted, nicely deduced from arguments

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+   const auto forward = direction == PlaybackDirection::forward;
+   while (clip)
+   {
+      const auto args = GetSampleAccessArgs<const float*>(
```
here too

Paul-Licameli commented:
`GetSamples` (not a new function) should also comment that `start` is relative to play start
`GetFloatAtTime` should comment that `t` is relative to play start

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+      const auto startTime =
+         std::max(startOrEndTime - clip.GetPlayStartTime(), 0.);
+      const sampleCount startSamp { std::ceil(
+         std::round(startTime * sampsPerSec * 1e6) / 1e6) };
```
Explain the strange use of `1e6`

Paul-Licameli replied:
Rounding to the nearest millionth of a sample?  Why?

saintmatthieu replied:
I can't remember. Just did a quick test with regular rounding, works too. I'll use a simple rounding, and next time should comment.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+      const auto len =
+         limitSampleBufferSize(remainingToRead, sampsInClip - startSamp);
+      return { reinterpret_cast<typename BufferCharType<BufferType>::type>(
+                  buffer + alreadyRead),
```
In the case that `max` above clipped a negative value to 0,  should the buffer instead get some zeroes at `buffer + alreadyRead` (if writing), and should the first member of the return be a later position?

Maybe not but I'm trying to understand the usage

saintmatthieu replied:
I'm not sure I understand your question.
`alreadyRead` means `[buffer, buffer + alreadyRead)` are indices that already have been dealt with.
`startSamp` tells which first sample index of the clip matches `buffer + alreadyRead`.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+      if (len == 0 || startSamp >= sampsInClip)
+         return { nullptr, sampleCount { 0u }, 0u };
+      const auto bufferEnd = buffer + remainingToRead;
+      return { reinterpret_cast<typename BufferCharType<BufferType>::type>(
```
Similar question whether this is the right thing when `min` above did some work

saintmatthieu replied:
If `min` did some work, then `startOrEndTime` was past the end of the clip.
> should the first member of the return be a later position?

It may be confusing to have this buffer offsetting amidst this more complex code. All the offset is doing is moving the read-write pointer according to the number of samples already read/written. It should not be concerned with the number of samples remaining to read from the clip.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+      if (numSamplesRead == numSamples)
+         break;
+      clip = GetAdjacentClip(*clip, direction);
+   }
```
This loop may concatenate samples from clips of different stretch ratios into one buffer, and then we forget about the stretches and operate on the buffer.

Understand that I objected to this for Plot Spectrum (weightings of windows in the combined result might be wrong) and for Mixer Board (RMS calculations may be wrong again for mis-weighting, but I let that problem go).

But here, doing microscopic repairs with your pencil or the even less-known paintbrush tool, it is defensible.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+            iChannel, args.offsetBuffer, floatSample, args.start, args.len,
+            effectiveFormat);
+         numSamplesWritten += args.len;
+         if (numSamplesWritten == numSamples)
```
I think `>=` makes me surer of loop termination just looking at this function.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+             mayThrow))
+         return 0u;
+      numSamplesRead += args.len;
+      if (numSamplesRead == numSamples)
```
Again `>=` would look better to me

saintmatthieu replied:
Ok (for both)

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+      numSamplesRead += args.len;
+      if (numSamplesRead == numSamples)
+         break;
+      clip = GetAdjacentClip(*clip, direction);
```
Here and elsewhere you use this or `GetNextClip`, there is a sorting of the clips each time.

Probably unimportant in practice for performance.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+      clip->SetFloatsFromTime(
+         tt0 - clipStartTime, iChannel, values.data(), numSamples,
+         effectiveFormat);
+      clip = GetNextClip(*clip, PlaybackDirection::forward);
```
Next, not Adjacent; and I understand why

Paul-Licameli commented on src/tracks/playabletrack/wavetrack/ui/SampleHandle.cpp:
```diff
@@ -142,7 +150,7 @@ UIHandlePtr SampleHandle::HitTest
    Envelope* env = wavetrack->GetEnvelopeAtTime(time);
    if (env)
       // Calculate sample as it would be rendered, so quantize time
-      envValue = env->GetValue( tt, 1.0 / wavetrack->GetRate() );
```
I think the "so quantize time" part no longer makes sense

Paul-Licameli commented on src/tracks/playabletrack/wavetrack/ui/SampleHandle.cpp:
```diff
@@ -290,9 +300,8 @@ UIHandle::Result SampleHandle::Click
       }
       //Set the sample to the point of the mouse event
```
This old comment should instead say, set a range of samples around the mouse event

Paul-Licameli commented on src/tracks/playabletrack/wavetrack/ui/SampleHandle.cpp:
```diff
@@ -345,66 +361,25 @@ UIHandle::Result SampleHandle::Drag
+   // updates, we draw a line going from v0 at t0 to v1 at t1.
+   const auto interpolator = [t0, t1, v0 = mLastDragSampleValue,
+                              v1 = newLevel](double t) {
+      const auto gradient = (v1 - v0) / (t1 - t0);
```
Does it ever divide by 0?

saintmatthieu replied:
Good question. The code before didn't check, but I cannot guarantee that my changes do not change that particular thing.
Adding a check upstream.

Paul-Licameli commented:
Done with this round of review!

I await a rebase onto #5138 and conflict resolution.

saintmatthieu commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
+         std::round(endTime * sampsPerSec * 1e6) / 1e6) };
+      const auto startSamp =
+         std::max(endSamp - remainingToRead, sampleCount { 0 });
+      const auto len = (endSamp - startSamp).as_size_t();
```
Will comment.

saintmatthieu commented:
@Paul-Licameli finished addressing your comments.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -198,6 +198,55 @@ void WaveClip::SetSamples(size_t ii,
+      return;
+   // Cannot be greater than `numFloats` -> safe cast
+   const auto numLeadingZeros =
+      (effectiveStart - maybeNegativeStart).as_size_t();
```
Actually, your informal proof here is wrong.  What if `t` happens to be a very negative value?

So to be safe you should write
`const auto numLeadingZeros = limitSampleBufferSize(numFloats, effectiveStart - maybeNegativeStart);`

saintmatthieu replied:
By substitution, `effectiveStart - maybeNegativeStart == std::max(sampleCount { 0 }, maybeNegativeStart) - maybeNegativeStart`,
and `maybeNegativeStart = TimeToSamples(t)`. So if `t < 0`, then `effectiveStart - maybeNegativeStart = - maybeNegativeStart`, that is, a positive value.

For `t` to be negative is actually not a corner case, here, and the code has proven to work well as far as my testing went, which is another evidence that it is correct.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.h:
```diff
@@ -326,6 +338,43 @@ class WAVE_TRACK_API WaveClip final :
+
+   //! Succeed with out-of-bounds requests, only changing what is in bounds.
+   //! @{
+   // clan-format off
```
You mean `clang-format` here and again below

saintmatthieu replied:
No, I meant `clan-format` XD

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.h:
```diff
@@ -326,6 +338,43 @@ class WAVE_TRACK_API WaveClip final :
+    * @brief Same as `SetFloatsFromTime`, but with `buffer` starting at
+    * `TimeToSamples(t0 -  SamplesToTime(numSideSamples))`.
+    * `[buffer, buffer + 2 * numSizeSamples + 1)` is assumed to be a valid span
+    * of addressed.
```
"addresses"

Paul-Licameli commented:
Finished this review round

Paul-Licameli reviewed (approved):
I'm still not really satisfied about the two uses of `as_size_t()`.  I'm dissatisfied now about the first, because,

`(effectiveEnd - effectiveStart)`

is bounded above by a `size_t` but there still isn't proof it is bounded below by 0 -- not negative.  You should first assign `effectiveEnd = std::max(effectiveStart, effectiveEnd);`

I am not persuaded either by the argument about the other `as_size_t()` and would rather see the `limitSampleBufferSize`.

I think then that without the other easy code changes, the function is still relying about unstated assumptions about its arguments so it has an incompletely stated contract.

I'd rather be deductive than argue that it works in practice so far.

But this PR was waited long enough.  I'm approving it, but with a protest.

saintmatthieu commented:
I was wrong about the first use: if `t` is past the end boundary of the clip, `effectiveEnd` can indeed be greater than `effectiveStart`. I couldn't reproduce this situation by trying even drag-drawing across several minuscule clips, but changing the value of `t` in the debugger showed it was possible. Thank you. Will push a safety check.

saintmatthieu commented:
Unit-testing would have been an easy way of demonstrating I was wrong.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -218,10 +218,9 @@ void WaveClip::SetFloatsFromTime(
-   if (effectiveLen == 0)
+   if (effectiveStart >= effectiveEnd)
       return;
+   const auto effectiveLen = (effectiveEnd - effectiveStart).as_size_t();
```
Keep the comment of old line 221

Paul-Licameli commented:
I see and approve the latest three commits.  Correct spelling of "rectangle" in the last commit comment.

But there are conflicts with master.

saintmatthieu commented:
Thanks for the speedy review.

dozzzzer commented:
Tested.
Found issues: #5231 
Also, I've discovered that when using the Pencil tool on adjacent clips, then stretching them and mixing and rendering the track, it produces a waveform that looks different from the pre-mixed-and-rendered state:


https://github.com/audacity/audacity/assets/70063970/ef7b7985-c065-4685-aefb-114d07f5c879


https://github.com/audacity/audacity/assets/70063970/5bea899e-e5bd-4690-aa6f-540e658fea20


https://github.com/audacity/audacity/assets/70063970/b7988d7e-1cc1-494a-9439-919869604cb0


Despite this mismatch doesn't look particularly good to me, this might be insignificant and I reckon should not delay the merge.
